axelor-supplychain/src/main/java/com/axelor/apps/supplychain/service/PurchaseOrderStockServiceImpl.java

```diff
@@ -320,35 +314,50 @@ protected StockLocation getStartStockLocation(PurchaseOrder purchaseOrder)
     return startLocation;
   }
 
-  protected Map<LocalDate, List<PurchaseOrderLine>> getAllPurchaseOrderLinePerDate(
+  protected Map<Pair<StockLocation, LocalDate>, List<PurchaseOrderLine>> getPurchaseOrderLineMap(
       PurchaseOrder purchaseOrder) {
 
-    Map<LocalDate, List<PurchaseOrderLine>> purchaseOrderLinePerDateMap = new HashMap<>();
+    Map<Pair<StockLocation, LocalDate>, List<PurchaseOrderLine>> purchaseOrderLineMap =
+        new HashMap<>();
 
     for (PurchaseOrderLine purchaseOrderLine : purchaseOrder.getPurchaseOrderLineList()) {
 
       if (purchaseOrderLineServiceSupplychainImpl.computeUndeliveredQty(purchaseOrderLine).signum()
           <= 0) {
         continue;
       }
+      StockLocation stockLocation = purchaseOrderLine.getStockLocation();
+      StockLocation purchaseOrderStockLocation = purchaseOrder.getStockLocation();
+
+      LocalDate dateKey =
+          Optional.ofNullable(purchaseOrderLine.getEstimatedReceiptDate())
+              .orElse(purchaseOrder.getEstimatedReceiptDate());
+
+      StockLocation stockLocationKey =
+          appStockService.getAppStock().getIsManageStockLocationOnStockMoveLine()
+                  && isStockLocationRelatedToPurchaseOrder(
+                      stockLocation, purchaseOrderStockLocation)
+              ? purchaseOrderStockLocation
+              : stockLocation;
+
+      purchaseOrderLineMap
+          .computeIfAbsent(Pair.of(stockLocationKey, dateKey), k -> new ArrayList<>())
+          .add(purchaseOrderLine);
+    }
 
-      LocalDate dateKey = purchaseOrderLine.getEstimatedReceiptDate();
-
-      if (dateKey == null) {
-        dateKey = purchaseOrderLine.getPurchaseOrder().getEstimatedReceiptDate();
-      }
-
-      List<PurchaseOrderLine> purchaseOrderLineLists = purchaseOrderLinePerDateMap.get(dateKey);
+    return purchaseOrderLineMap;
+  }
 
-      if (purchaseOrderLineLists == null) {
-        purchaseOrderLineLists = new ArrayList<>();
-        purchaseOrderLinePerDateMap.put(dateKey, purchaseOrderLineLists);
+  protected boolean isStockLocationRelatedToPurchaseOrder(
+      StockLocation stockLocation, StockLocation purchaseOrderStockLocation) {
+    StockLocation currentLocation = stockLocation;
+    while (currentLocation != null) {
```

Comment:
please add a counter to prevent infinite loop, just to be sure (could happen in case of wrong stock location import)